cli/report/report.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package report
+
+import (
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/open-policy-agent/opa/loader"
+	"github.com/open-policy-agent/opa/rego"
+)
+
+// GenerateReports takes CAI output and generate inventory report content based on report queries
+func GenerateReports(caiPath string, reportQueryPath string, reportOutputPath string, reportFormat string) error {
+	fileSuffix := time.Now().Format("20060102150405")
+	rawAssetFileName, err := convertAndGenerateTempAssetFile(caiPath, reportOutputPath, fileSuffix)
+	if err != nil {
+		return err
+	}
+	results, err := generateReportData(rawAssetFileName, reportQueryPath, reportOutputPath)
+	printReports(results, reportOutputPath, reportFormat, fileSuffix)
+	return nil
+}
+
+func convertAndGenerateTempAssetFile(caiPath string, reportOutputPath string, fileMidName string) (rawAssetFileName string, err error) {
+	results, err := ReadFilesAndConcat(caiPath)
+	if err != nil {
+		return "", err
+	}
+	wrapped := map[string]interface{}{
+		"assets": results,
+	}
+	outJSON, _ := json.MarshalIndent(wrapped, "", "  ")
+	rawAssetFileName = "raw_assets_" + fileMidName + ".json"
+	err = ioutil.WriteFile(filepath.Join(reportOutputPath, rawAssetFileName), outJSON, 0644)
+	if err != nil {
+		return "", err
+	}
+	return
+}
+
+func generateReportData(rawAssetFileName string, reportQueryPath string, reportOutputPath string) (results interface{}, err error) {
+	// Load resources from json and rego files
+	resources, err := loader.All([]string{filepath.Join(reportOutputPath, rawAssetFileName), reportQueryPath})
+	if err != nil {
+		return nil, err
+	}
+	compiler, err := resources.Compiler()
+	if err != nil {
+		return nil, err
+	}
+	store, err := resources.Store()
+	if err != nil {
+		return nil, err
+	}
+	r := rego.New(
+		rego.Query(`data.reports`),
+		rego.Compiler(compiler),
+		rego.Store(store),
+	)
+	rs, err := r.Eval(context.Background())
+	if err != nil {
+		return nil, err
+	}
+	results = rs[0].Expressions[0].Value
+
+	return results, nil
+}
+
+func printReports(results interface{}, reportOutputPath string, format string, fileSuffix string) error {
+	resultsMap := results.(map[string]interface{})
+	for group, reports := range resultsMap {
+		reportsMap := reports.(map[string]interface{})
```

Comment:
These aren't all reports though? Maybe consider a more descriptive name. `groupItems` or something generic?

Reply:
As of now these things are all inventory reports
A "group" can be "iam", "vm", "network", "data" etc,
and a "report" can be "iam.bindings_report", "vm.service_account_report", "network.subnet_flow_logs_report", "data.cloud_sql_public_authorized_networks_report", etc

Reply:
Got it.